### DISCUSSION TITLE
RFC: Integrate SVGR (or equivalent functionality)

### DISCUSSION THREAD

**Discussion by kylemh** [vercel/next.js#26167] [Ideas]
# Edited

It's currently trivial to integrate SVGR thanks to https://github.com/vercel/next.js/pull/26281; however, we still lose out on the ability to choose to use an SVG with next-images OR SVGR. We get one or the other, but not both.

I believe Next should incorporate SVGR (or at least the ability to do a named import of an SVG and have it be a component) as either a plugin or the core webpack package. This would enable the prime SVGR feature without sacrificing the ability to use SVGs with next-image.

[SVGR has long supported the ability to peacefully coexist with file-loader and url-loader usages](https://react-svgr.com/docs/webpack/) by using named imports. So, there exists a way to integrate this without a breaking change and in keeping with standard practice (from both SVGR's "API" and Next's).

```js
import starUrl, { ReactComponent as Star } from './star.svg'
const App = () => (
  <div>
    <img src={starUrl} alt="star" />
    <Star />
  </div>
)
```

# This is how this issue was opened initially

### Describe the feature you'd like to request

I would like the ability to import SVGs and use them directly as components. This is useful for iconography and complex SVG animations, because the SVG will end up as a React component somehow, someway anyways.

### Describe the solution you'd like

Integrate SVGR directly into Next.js's webpack configuration and support the ability to import an SVG directly as a component and update the type definitions for `'*.svg'` imports.

[SVGR has long supported the ability to peacefully coexist with file-loader and url-loader usages](https://react-svgr.com/docs/webpack/) by using named imports. So, there exists a way to integrate this without a breaking change and in keeping with standard practice.

```js
import starUrl, { ReactComponent as Star } from './star.svg'
const App = () => (
  <div>
    <img src={starUrl} alt="star" />
    <Star />
  </div>
)
```

### Describe alternatives you've considered

Today, I'm able to use https://github.com/newhighsco/next-plugin-svgr (and there many other simple ways to customize the webpack configuration to leverage the tool. You're able to now, but then you won't be able to use SVGs in next/image.

**Comment by kylemh** [comment]
I'd love to contribute this myself too. Let me know if there's a reason to NOT open a PR for this.

**Reply by timneutkens** [reply]
For one the package is way too large: over 6.5MB install size https://packagephobia.com/result?p=@svgr/core so that'll have to be significantly reduced in order to add it to Next.js to begin with (it's a 15% increase in install size). Other significant downsides are that importing svgs in that way causes them to be inlined in the JS bundle which can significantly increase bundle size when used incorrectly. For example we've seen cases where complete icon libraries imported in this was are included in the JS bundle even though only a few icons are used in practice, that is one of the reasons it's not provided by default.

**Reply by kylemh** [reply]
With route-based code splitting, wouldn’t the 2nd problem you listed not be an issue?

**Reply by kylemh** [reply]
Regarding your first problem, that bundle size increase wouldn’t come back to the consumer. Why is it such a big priority to keep devDependencies small?

**Reply by timneutkens** [reply]
> With route-based code splitting, wouldn’t the 2nd problem you listed not be an issue?

Nope, what we've seen was on Next.js apps that manually added the approach you're suggesting.

> Why is it such a big priority to keep devDependencies small?

It significantly slows down `npm install`, especially when the size increase is not needed at all, for example the plugin you're suggesting is shipping it's own full version of Babel from the looks of it, that should use the internal version Next.js has which will likely make it a lot smaller. 

If we did not prioritize package size we'd now have about 4 to 5x larger install size.

**Reply by kylemh** [reply]
Okay, I'm hearing that SVGR isn't currently up to snuff. That is potentially fine because we can just try to implement the feature without the dependency; however, I'm also still worried about what you're saying here, because it implies the feature is actively something you don't want:

> Nope, what we've seen was on Next.js apps that manually added the approach you're suggesting.

Still confused by this. Presuming SVGR was integrated, and if I do:
```tsx
import { ReactComponent as CommentIcon } from 'public/svgs/icons/comment-icon.svg';
```

on some component. What makes it so the bundle increases unnecessarily? I would think that the bundle would increase for only the pages using this component.

**Reply by timneutkens** [reply]
Because people end up doing stuff like this:

```
export default {
 icon1: CommentIcon,
 icon2: CommentIcon2
 icon3: CommentIcon3
}
```

and other interesting stuff that can't be tree shaken. It'll likely have to be very strict if integrated into Next.js.

**Reply by kylemh** [reply]
Understood. Thanks for taking the time to help me understand.

I feel comfortable at least attempting the babel/webpack piece; however, the missing piece to make you content with the added feature sounds like it should lean on conformance. I'm unfamiliar with making lint rules, especially ones that need to be complex enough to look for foot-guns like the one you just outlined. Is it enough to begin working on the feature set without the actual SVGR dependency?

I would implement this _without_ SVGO considering it also has a large install size; however, now that I understand your concerns fully... I wonder if an official plugin is allowed more leeway in install size? If so, I think an official plugin + conformance tactics would be a good way to assuage any issues. Rest assured, if plugins are allowed more leeway it will only be to add SVGO - not to install all the babel deps.

**Reply by MaximeBernard** [reply]
"Because people can do silly stuff with your tool" should not be a reason not to provide the tool. I'm quite confused with that reasoning.

> If we did not prioritize package size we'd now have about 4 to 5x larger install size.

Even though it's always nice to have a quick install, I'd rather have a nice feature than a quick install. (we're talking about 6.5MB which will take me 2~3 seconds more the first time and then it the cache)

**Reply by timneutkens** [reply]
@MaximeBernard I've never said we won't add it, I've laid out the reasoning for why it was not added yet and what issues to solve before it can be added, 2 are very clear:

- install size
- conformance around ensuring that you don't include hundreds of kbs of svgs into pages

I've even given the solution to install size: svgr ships it's own version of Babel which is not needed in the context of Next.js. Removing that shaves off 6.31MB of the install size when used in Next.js.

>  I'd rather have a nice feature than a quick install.

While I get you want this feature and you want it now, this is not how we approach long-term maintainability and ensuring Next.js stays fast and gets faster over time.

**Reply by kylemh** [reply]
Thoughts on plug-in with SVGO vs core without SVGO?

Judging by another thread, people can use a separate SVGO loader and it appears to work in v11+

**Reply by MaximeBernard** [reply]
Thank you for the clarification @timneutkens 👍

**Reply by kylemh** [reply]
necroing this one... is tree-shakeability the main concern or the underlying bundle... or both? i'm curious if SVGR-like functionality would be acceptable alongside `use` cacheing: https://paco.me/writing/svg-caching-with-use

as an aside: thinking out loud here and maybe missing bad problems... perhaps there's something that can be done 1st-class within Next.js where there's an ideal folder for .svg iconography or something

**Comment by njmh** [comment]
This is a must IMO - or some other native solution to inlining SVGs if SVGR is deemed too big.  Whatever the solution is and as much as I appreciate the work they do, declaring "*.svg" in the global.d.ts as StaticImageData was far from a good choice by the Vercel team.

**Comment by kylemh** [comment]
Another important aspect that a friend made me aware of. Without internalizing SVGR, there's no way to advertise that the CRA codemod will finish migrating to Next.js if its feature set is in use.

**Comment by kylemh** [comment]
https://github.com/vercel/next.js/pull/26281 was merged into `v11.0.1-canary.4` which will make most people happy. I'll update OP, but I still think integrating SVGR is a good idea. In summary, it would still be cool to take advantage of next-images for SVGs when doing a default import, and have the ability to do a component import as a named import.

**Comment by flybayer** [comment]
I 100% agree we should bake this into core!

**Comment by pcjmfranken** [comment]
The SVGR package doesn't seem very well maintained. Several issues and peer-reviewed PRs that fix problems with Webpack 5 compatibility and the 5 high severity vulnerabilities that NPM currently reports with SVGR seem to be ignored by the maintainers.

Another thing to take into consideration would be how to go about custom configuration. It's very common to have a separate config targeting `*.icon.svg` files that optimises them for use as (wait for it.. 😁) icons (this could be implemented as a default or flag).

The core concept would be a very welcome addition to Next.js, though. I think it also fits perfectly with its `No config needed.` mission.

**Reply by gregberge** [reply]
Sorry I got a kid, and no donation to work on my paid hours.

**Reply by pcjmfranken** [reply]
Hey @gregberge, that wasn't in any way meant to blame or otherwise accuse you of anything. I'm truly terribly sorry if you picked it up that way!

**Reply by gregberge** [reply]
No problem! The project is now well maintained, I managed to got time to do it!

**Comment by jserrao** [comment]
If SVGR doesn't get integrated into Next, what is the recommended way to handle favicons? Do we somehow run that through `next/image` as well?

**Reply by kylemh** [reply]
Favicons specifically are a different thing, no? Does SVGR somehow convert svg files to ico files? You can handle favicons with [static file serving](https://nextjs.org/docs/basic-features/static-file-serving).

**Reply by jserrao** [reply]
I like to use SVG favicons but maybe that's just not possible

**Reply by kylemh** [reply]
Well, then you could integrate SVGR yourself and use SVG favicons. [Keep in mind SVG favicons are still not universally supported](https://caniuse.com/link-icon-svg).

**Reply by jserrao** [reply]
It's kind of funny, Apple dictates like 800 favicons for every project but hasn't prioritized supporting SVG favicons in Safari. 

My larger point is that you can't use next/image with SVG, PNG, etc for favicons. They require a direct SVG/PNG import that isn't possible right now - try it out, you'll see what I mean.

Next is doing some voodoo under the hood that lets you place a favicon.ico in the `/public` folder and, without importing, it just works with their custom builder in 12.x+. But, as you point out, there are a lot of legacy favicon formats and no good way to use them with the new build tool. Seems crazy to put loaders into the mix just to support favicons, no?

**Reply by alebelcor** [reply]
@jserrao Follow [this](https://evilmartians.com/chronicles/how-to-favicon-in-2021-six-files-that-fit-most-needs) and you should be fine.

No need to use `next/image`.
